### ISSUE TITLE
doesn't support additionalProperties: true

### ISSUE THREAD

**Issue by ysulyma** [x52dev/oas3-rs#52]
Spec
```yaml
openapi: 3.0.0
info:
  version: 1.0.0
  title: Swagger Petstore
components:
  schemas:
    Pet:
      type: object
      additionalProperties: true
paths:
  /pet:
    get:
      responses:
        204:
          description: Ok
```
Rust code
```rust
use oas3;

fn main() {
    match oas3::from_path("./petstore-minimal.yaml") {
        Ok(spec) => println!("spec: {:#?}", spec),
        Err(err) => println!("error: {:?}", err),
    }
}

```
Output
```
error: Yaml(Error("components.schemas: data did not match any variant of untagged enum ObjectOrReference", line: 7, column: 5))
```

**Comment by dahrens** [comment]
The openapi spec is not clear about `addionalProperties`, there is an [open issue](https://github.com/OAI/OpenAPI-Specification/issues/3720) that addresses the boolean value.
Right now it defines what oas-rs currently parses (but not honors on validation) - passing objects or refs.

The [swagger documentation](https://swagger.io/docs/specification/data-models/data-types/) meanwhile has defined it, see "Free-Form Object". They also assume that when `additionalProperties` is omitted, it defaults to `true`.

> A free-form object (arbitrary property/value pairs) is defined as:
>
> ```
>    type: object
> ```
>
> This is equivalent to
> 
> ```
>    type: object
>    additionalProperties: true
> ```
>
> and
>
> ```
>    type: object
>    additionalProperties: {}
> ```

**Comment by robjtede** [comment]
I've commited a change that clarifies this crate follows OAS v3.1, not v3.0

**Comment by dahrens** [comment]
The new version supports parsing a schema that uses `additionalProperties: true`, but it does not validate properly in that case as it always assumes that `additionalProperties` is false.

This [commit](https://github.com/dahrens/oas3-rs/commit/7c61d95dfc5e384140bdae74f1988fd16684d72f) illustates the issue.

**Comment by robjtede** [comment]
From the OAS spec:

> Unless stated otherwise, the property definitions follow those of JSON Schema and do not add any additional semantics.

From the 2020 JSON schema draft:

> The value of "additionalProperties" MUST be a valid JSON Schema.

I'm struggling to find anywhere that states a v3.1 spec supports `additionalProperties: false`. It is only [stated in v3.0](https://github.com/OAI/OpenAPI-Specification/blob/main/versions/3.0.3.md#properties).

**Comment by robjtede** [comment]
Apoligies if I'm missing something obvious. Actually I notice now your commit link goes to the validation code. If this is what you mean by not supporting `additionaProperties` then I understand, the validation code is not up to spec yet (hence the feature flag).

**Comment by simfeld** [comment]
@robjtede I ran into the same issue of `additionalProperties: false` leading to a parsing error. My take on the 3.1 spec is, as you said
> Unless stated otherwise, the property definitions follow those of JSON Schema and do not add any additional semantics.

and the current json-schema draft says the following about `additionalProperties` (https://json-schema.org/draft/2020-12/draft-bhutton-json-schema-01#name-additionalproperties):
> The value of "additionalProperties" MUST be a valid JSON Schema.

So the question is whether a boolean is a valid JSON schema, and I think it is:
https://json-schema.org/draft/2020-12/draft-bhutton-json-schema-01#name-json-schema-documents says
> A JSON Schema MUST be an object or a boolean.

https://json-schema.org/draft/2020-12/draft-bhutton-json-schema-01#name-boolean-json-schemas says
> The boolean schema values "true" and "false" are trivial schemas that always produce themselves as assertion results, regardless of the instance value.

and on top of that https://json-schema.org/understanding-json-schema/reference/object#additionalproperties mentions the usage of false as well:
> Setting the additionalProperties schema to false means no additional properties will be allowed.

WDYT?

**Comment by robjtede** [comment]
Awesome, thanks so much for the spec references. I'd definitely not dug deep enough into the JSON schema spec.

**Comment by robjtede** [comment]
fixed in 81d711146b7e625a0d7a4d303e0c43276a648e88

**Comment by robjtede** [comment]
released in v0.9.0

**Comment by simfeld** [comment]
thanks for the quick fix @robjtede, awesome! I can confirm that v0.9 now parses `additionalProperties: false`
